### ISSUE TITLE
[FEAT] Parallelize Checks

### ISSUE THREAD

**Issue by ShrinathVis** [deepchecks/deepchecks#1576]
Hi, love the library! Produces a lot of great information.

One thing I noticed was that some of the checks take a while, especially on a large train dataset with a large number of features. I was wondering if we could add a parallelization across different checks in Suite.run() using multiprocessing?

Thanks!

**Comment by ItayGabbay** [comment]
@ShrinathVis Thanks! I agree it's worth trying :) 
It will be really helpful if you can specify what are the dimensions of your dataset and how much time it takes to run to run the suite (and which suite). 

@matanper @yromanyshyn @JKL98ISR I wonder what are your thoughts on this?

**Comment by matanper** [comment]
We might need to do some work on the `Context` object to make it thread-safe

**Comment by yromanyshyn** [comment]
@ItayGabbay 
yup, it is definitely worth dedicating some time to experiment with that to find cases when using multiprocessing is profitable

but using multiprocessing for suites with few checks or a small amount of data can produce totally opposite result, you know, processes creation and data sharing between them`s are not cheap

plus, suite/checks will require some modifications, to make it possible to use multiprocessing, and currently, it is hard to say what kind of modifications and to what extend

it all requires exploration and experimenting

> We might need to do some work on the Context object to make it thread-safe (for example add a lock on the feature importance calculation)

thread-safety is not applicable to the multiprocessing because you do not have shared mutable memory,

**Comment by matanper** [comment]
@yromanyshyn there are specific data types that can be shared across processes, my idea was to use them in order to prevent things like calculating the FI for each check separately.
 https://docs.python.org/3/library/multiprocessing.shared_memory.html


> but using multiprocessing for suites with few checks or a small amount of data can produce totally opposite result
In theory you are correct, but in practice we do not care about second-precision run times. If for small dataset it will take 35 instead of 30 seconds it doesn't matter much, but if for larger dataset it will run in 5 minutes instead of 60 minutes then that's a big improvement
